node/src/components/linear_chain_sync/operations.rs

```diff
@@ -455,7 +493,7 @@ pub(crate) async fn run_fast_sync_task(
         .era_start
         .saturating_diff(current_header.timestamp())
         < chainspec.deploy_config.max_ttl
-        && !current_header.is_genesis_child()
+        && current_header.height() != 0
     {
         current_header =
             fetch_and_store_block_by_hash(effect_builder, *current_header.parent_hash())
```

Comment:
I just realized that this won't even check storage — it will always download the block again, even if it's already there? We should skip the blocks that we already have. (But if we only have the header, we do need to download the body, of course.)

Reply:
I think it will - it uses `effect_builder.fetch` internally, which calls to `Fetcher` which in turns does check the storage firstly. The problem here is that `fetch_retry_forever` also calls `put_block_to_storage` which shouldn't be necessary if we found that block in the storage in the first place.

Reply:
Should be fixed in afa4a647170b3c161cfa706919d1f5132c8ea7b6